src/Kestrel.Core/Internal/Infrastructure/TimeoutControl.cs

```diff
@@ -124,6 +124,13 @@ private void CheckForWriteDataRateTimeout(long timestamp)
         {
             lock (_writeTimingLock)
             {
+                var extraTimeForTick = timestamp - _lastTimestamp - Heartbeat.Interval.Ticks;
```

Comment:
I was curious whether this could potentially be racy if multiple ticks end up running concurrently but I think it might be rare enough that it's not a problem. Essentially, we expect the _lastTimestamp to be updated after this value is calculated. However, there's no guarantee that CheckForWriteDataRateTimeout and Interlocked.Exchange(ref _lastTimestamp, timestamp) run atomically so concurrent ticks may add extra extraTimeForTick. Not only would this be rare, the effect would be that we allow a little extra time for write, which doesn't seem too bad.

Reply:
Heartbeat.cs prevents overlapped ticks, so there shouldn't be a race here.